runtime.md

```diff
@@ -29,24 +29,60 @@ The root directory to the bundle is provided in the state so that consumers can
 }
 ```
 
-## Lifecycle
+## Typical lifecycle
+
+A typical lifecyle progresses like this:
+
+1. There is no container
+2. A user tells the runtime to start a container and launch a process inside it
+3. The runtime [creates the container](#create)
+4. The runtime executes any [pre-start hooks](runtime-config.md#pre-start)
+5. The runtime [executes the container process](#start-process)
+6. The container process is running
+7. The runtime executes any [post-start hooks](runtime-config.md#post-start)
+8. A user tells the runtime to send a termination signal to the container process
+9. The runtime [sends a termination signal to the container process](#stop-process)
+10. The container process exits
+11. The runtime [terminates any other processes in the container](#stop-process)
+12. The runtime executes any [post-stop hooks](runtime-config.md#post-stop)
+13. The runtime [removes the container](#cleanup)
+
+With steps 7 and 8, the user is explicitly stopping the container process (via the runtime), but it's also possible that the container process could exit for other reasons.
+In that case we skip directly from 6 to [10](#stop-process), skipping any post-start hooks that hadn't been launched and terminating any in-progress post-start hook.
+
+Failure in a pre-start hook or other setup task can cause a jump straight to [12](runtime-config.md#post-stop).
 
 ### Create
 
-Creates the container: file system, namespaces, cgroups, capabilities.
+Create the container: file system, namespaces, cgroups, capabilities, etc.
+The invoked process forks, with one branch that stays in the host namespace and another that enters the container.
+The host process carries out all container setup actions, and continues running for the life of the container so it can perform teardown after the container process exits.
+The container process changes users and drops privileges in preparation for the container process start.
+At this point, the host process writes the [`state.json`](#state) file with the host-side version of the container-process's PID (the container process may be in a PID namespace).
 
 ### Start (process)
 
-Runs a process in a container.
-Can be invoked several times.
+After the pre-start hooks complete, the host process signals the container process to execute the runtime.
```

Comment:
On Wed, Oct 07, 2015 at 11:51:50AM -0700, W. Trevor King wrote:
> Wed, Oct 07, 2015 at 11:40:34AM -0700, Michael Crosby:
> > > +After the pre-start hooks complete, the host process signals
> > > the container process to execute the runtime.
> >
> > This sentence does not make sense.  ` the host process signals the
> > container process to execute the runtime.`????
> 
> I haven't looked into the runC implementation here, but my guess was
> that:
> 
> …
> 6. The host process would signal the container process that the
>    pre-start hooks had completed successfully.
> 7. The container process would catch that signal and exec the
>    config.json process.
> 
> The line you're quoting was supposed to sketch out (6) and (7).  Can
> you clear me up on how it actually works?

It looks like libcontainer actually has a fairly detailed spec for how
it sets up containers.  That talks about using a pipe instead of
singals for this trigger [1].  Signals sound easier to me (permissions
discussed in [2]), but the generic “signals” verb could cover all such
triggers (including piped messages) and not be restricted to POSIX
signals [3], since this is an internal runtime mechanism.

[1]: https://github.com/opencontainers/runc/blob/v0.0.4/libcontainer/SPEC.md#runtime-and-init-process
[2]: http://man7.org/linux/man-pages/man2/kill.2.html
[3]: http://man7.org/linux/man-pages/man7/signal.7.html